### PR TITLE
Validate rules ontologically

### PR DESCRIPTION
**Pull request by kasper-piskorski**

### CODE REVIEW

**Comment by kasper-piskorski** [comment]
please retest this please - grakn-dist failure

File: grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java [hunk 1]
```diff
@@ -79,6 +80,12 @@
     default boolean isAllowedToFormRuleHead(){ return false; }
 
     /**
+     * @return error messages indicating ontological inconsistencies of this atomic
+     */
+    @CheckReturnValue
+    default Set<String> validateOntologically(){ return new HashSet<>();}
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java]: Why do you need a default which returns an empty set? Just curious.

File: grakn-core/src/main/java/ai/grakn/util/ErrorMessage.java [hunk 2]
```diff
@@ -96,6 +96,15 @@
 
     VALIDATION_RULE_ILLEGAL_ATOMIC_IN_HEAD("The rule [%s] of type [%s] does not form a valid Horn clause, as its head contains illegal atomics\n"),
 
+    VALIDATION_RULE_INVALID_RELATION_TYPE("Attempting to define a rule containing a relation pattern with type [%s] which is not a relation type\n"),
+
+    VALIDATION_RULE_INVALID_RESOURCE_TYPE("Attempting to define a rule containing a resource pattern with type [%s] which is not a resource type\n"),
+
+    VALIDATION_RULE_RESOURCE_OWNER_CANNOT_HAVE_RESOURCE("Attempting to define a rule containing a resource pattern of type [%s] with owner type [%s] that cannot have this resource\n"),
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/util/ErrorMessage.java]: maybe just change `owner type` to `type`. `owner type` makes me think there is another type out there.

File: grakn-core/src/main/java/ai/grakn/util/ErrorMessage.java [hunk 3]
```diff
@@ -96,6 +96,15 @@
...
+
+    VALIDATION_RULE_ROLE_CANNOT_BE_PLAYED("Attempting to define a rule containing a relation pattern with role [%s] which is cannot be played in relation [%s]\n"),
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/util/ErrorMessage.java]: Grammer Nitpick:

> Attempting to define a rule containing a relation pattern with role [%s] which ~is~ cannot be played in relation [%s]\n

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java [hunk 4]
```diff
@@ -285,20 +285,41 @@ private ValidateGlobalRules() {
      * @param rule the rule to be validated
      * @return Error messages if the rule is not a valid Horn clause (in implication form, conjunction in the body, single-atom conjunction in the head)
      */
-    static Set<String> validateRuleIsHornClause(GraknGraph graph, Rule rule){
+    static Set<String> validateRuleIsValidHornClause(GraknGraph graph, Rule rule){
         Set<String> errors = new HashSet<>();
-        if (rule.getWhen().admin().isDisjunction()){
+        if (rule.getWhen().admin().getDisjunctiveNormalForm().getPatterns().size() != 1){
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java]: I kind of preferred `isDisjunction` much cleaner and more readable. Why did you have to change it?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java [hunk 5]
```diff
@@ -285,20 +285,41 @@ private ValidateGlobalRules() {
...
             errors.add(ErrorMessage.VALIDATION_RULE_DISJUNCTION_IN_BODY.getMessage(rule.getId(), rule.type().getLabel()));
         }
         errors.addAll(checkRuleHeadInvalid(graph, rule, rule.getThen()));
         return errors;
     }
 
     /**
+     * NB: assumes the rule obeys the Horn clause form
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java]: Rephrase to say that this only gets checked if the rule is a valid horn clause.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java [hunk 6]
```diff
@@ -285,20 +285,41 @@ private ValidateGlobalRules() {
...
+     * @param graph graph used to ensure the rule is a valid Horn clause
+     * @param rule the rule to be validated ontologically
+     * @return Error messages if the rule has ontological inconsistencies
+     */
+    static Set<String> validateRuleOntologically(GraknGraph graph, Rule rule) {
+        Set<String> errors = new HashSet<>();
+
+        //both body and head refer to the same graph and have to be valid with respect to the ontology that governs it
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ValidateGlobalRules.java]: Nice :)

File: grakn-graql/src/test/java/ai/grakn/graph/internal/RuleTest.java [hunk 7]
```diff
@@ -226,21 +226,108 @@ public void whenAddingRuleWithIllegalAtomicInHead_PropertyAtoms_Throw() throws I
...
-        Role role3 = graknGraph.putRole("role3");
 
-        graknGraph.putRelationType("relation1")
-                .relates(role1)
-                .relates(role2)
-                .relates(role3);
-        graknGraph.putRelationType("relation2")
-                .relates(role3);
+    @Test
+    public void whenAddingRuleInvalidOntologically_RelationDoesntRelateARole_Throw() throws InvalidGraphException{
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graph/internal/RuleTest.java]: 👍

File: grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java (hunk 1 above)
Comment (kasper-piskorski) [inline grakn-core/src/main/java/ai/grakn/graql/admin/Atomic.java]: It makes an atomic ontologically valid by default. I assume only resources and relations can be potentially invalid.

**Pull request merged**